algorithms/Matrix-factorization/Recom/camf_c_movielens.py

```diff
@@ -0,0 +1,185 @@
+                + self.average_item_ratings[row['movieId']]
+                + self.user_biases[user_index]
+                + self.time_of_day_biases[time_of_day_index]
+                + self.day_of_week_biases[day_of_week_index])
+            predictions[row['movieId']] = (rating, predicted_rating)
+        return predictions
+
+    def train(self, training_data):
+        """
+        Function that trains the model by learning the parameters
+        through stochastic gradient descent
+        """
+
+        lrate = 0.001
+        rterm = 0.001
+        for epoch in range(1000):
+            batch = training_data.sample(ceil(len(training_data.index)/1000))
+
+            for _, row in batch.iterrows():
+                user_index = self.users_dict[row['userId']]
+                item_index = self.items_dict[row['movieId']]
+                time_of_day_index = self.time_of_day_dict[row['timeofday']]
+                day_of_week_index = self.day_of_week_dict[row['dayofweek']]
+                rating = row['rating']
+
+                error = rating - (np.dot(
+                    self.user_factor_matrix[user_index, :],
+                    self.factor_item_matrix[:, item_index])
+                    + self.average_item_ratings[row['movieId']]
+                    + self.user_biases[user_index]
+                    + self.time_of_day_biases[time_of_day_index]
+                    + self.day_of_week_biases[day_of_week_index])
+
+                self.user_biases[user_index] = self.user_biases[
+                    user_index] - lrate * (2 * error * (-1) +
+                                           2 * rterm *
+                                           self.user_biases[user_index])
+                self.time_of_day_biases[time_of_day_index] = \
+                    self.time_of_day_biases[time_of_day_index] - \
+                    lrate * (2 * error * (-1) + 2 * rterm *
+                             self.time_of_day_biases[time_of_day_index])
+                self.day_of_week_biases[day_of_week_index] = \
+                    self.day_of_week_biases[day_of_week_index] - \
+                    lrate * (2 * error * (-1) + 2 * rterm *
+                             self.day_of_week_biases[day_of_week_index])
+
+                for k in range(self.factors):
+                    self.user_factor_matrix[user_index][k] = \
+                        self.user_factor_matrix[user_index][k] - lrate * \
+                        (2 * error * (-1) *
+                         self.factor_item_matrix[k][item_index]
+                         + 2*rterm * self.user_factor_matrix[user_index][k])
+                    self.factor_item_matrix[k][item_index] = \
+                        self.factor_item_matrix[k][item_index] - lrate * (
+                        2 * error * (-1) *
+                        self.user_factor_matrix[user_index][k] +
+                        2 * rterm * self.factor_item_matrix[k][item_index])
+
+            if epoch % 10 == 0:
```

Comment:
HVad er forskellen på de her udregner af mse og mae of dem i caluclate metrics functionen?

Reply:
Det bruger den samme funktion nu :)